BB-444: Revision Display Page Improvement (WIP)

prabalsingh24 opened the pull request:
WIP
PR not ready for review rn

prabalsingh24 commented:
This isn't ready for a review right now. 

What other changes should be made @MonkeyDo ?

![Screenshot from 2020-04-10 15-43-34](https://user-images.githubusercontent.com/45737735/78983743-69203d80-7b42-11ea-92b7-b9e4b372f0ac.png)

I think the way relationship is shown here is not very good. I added four relationships to "God Emperor of Dune" and the bbids in the table are of this work itself. I think it would be better if those bbids were of the other entity whom which this work is related to.

Even though "Public Affairs" and "Wonder Woman" were revised, their changes are not shown ( a bug maybe).

What do you think?

chinmay-kothari commented:
@prabalsingh24 
Writing AliasName there may not look good when we have long AliasNames like: https://beta.bookbrainz.org/edition/4e4788aa-b139-47e3-8674-dad226ddc9aa

What do you think?

MonkeyDo commented:
It definitely looks better. Maybe it would be good to also show the BBID? maybe under the title in a `<small>` tag?

I agree the way relationships are displayed is terrible. There's lots of low hanging fruits to make that page better.
To display relationships, in an ideal world I would love something like the diagram in the [help page](https://bookbrainz.org/help), but that might be for another day.

As for @chinmay-kothari 's concern about long names, you'll have to make sure the name wraps correctly, but I don't think we have many other options.

MonkeyDo commented:
> Even though "Public Affairs" and "Wonder Woman" were revised, their changes are not shown ( a bug maybe).

Possibly. How did you make changes to the relationships of God Emperor of Dune as well as Public Affairs and Wonder Woman at the same time?
In any case, it looks like it would be worth investigating.

prabalsingh24 commented:
> > Even though "Public Affairs" and "Wonder Woman" were revised, their changes are not shown ( a bug maybe).
> 
> Possibly. How did you make changes to the relationships of God Emperor of Dune as well as Public Affairs and Wonder Woman at the same time?
> In any case, it looks like it would be worth investigating.

![Screenshot_2020-04-10 BookBrainz – The Open Book Database](https://user-images.githubusercontent.com/45737735/78999463-502a8300-7b68-11ea-8ac7-e14174c66d69.png)

I think it's a bug. Here in this revision. I added the relationship - "New Author" wrote "New Work2" and it didn't show up in both. 
I am having a look

prabalsingh24 commented:
I was editing "God Emperor of Dune" and I kept adding relationships.

prabalsingh24 commented:
`formatNewRelationshipSet` function will be used when you add relationships to an entity w/o any previous relationships.

prabalsingh24 commented:
The `formatRemoveRelationships` function isn't perfect yet...
It's having bug in following case.
Let's say the old relationship set is having following `ids``[100 , 200, 300, 400, 500, 600]` and the new relationship set is `[100, 200, 400, 500, 600]` 
Deep-Diff library which is used to find the changes will show that elements at 2nd, 3rd, 4th index are edited. and the element in 5th index (600) is deleted but in reality element 300 was deleted..

The `formatAddRelationships` is working fine because every time a new relationship is created, it's relationshipId will always greater than the older ids (I think?) and because of it, it will always be at the last index ( because we are sorting before finding the diffs)

prabalsingh24 commented:
![Screenshot_2020-04-13 BookBrainz – The Open Book Database](https://user-images.githubusercontent.com/45737735/79108116-49cc1f00-7d93-11ea-8db5-7b2df2a7c44d.png)

Instead of showing bbid of the entity itself, I changed it to show bbid of the other entity. the one it's related to. (check table)
I made a mistake in column name, I am gonna fix it in next commit. Don't wanna create an extra commit for a typo :(

prabalsingh24 commented:
@MonkeyDo Tagging you because the option to ask for a review is not showing up :P

prabalsingh24 commented:
![Screenshot_2020-04-16 BookBrainz – The Open Book Database](https://user-images.githubusercontent.com/45737735/79429956-d79f4880-7fe5-11ea-8c0a-8efbcc3bf959.png)



In this revision, I removed `MusicBrainz ID`, but this shows that I removed `WikiData` and then 
added `WikiData` inplace of `MusicBrainz ID`.

`Deep-Diff` isn't working properly. Should it be replaced w some different library?

MonkeyDo commented:
> ![Screenshot_2020-04-16 BookBrainz – The Open Book Database](https://user-images.githubusercontent.com/45737735/79429956-d79f4880-7fe5-11ea-8c0a-8efbcc3bf959.png)
> 
> In this revision, I removed `MusicBrainz ID`, but this shows that I removed `WikiData` and then
> added `WikiData` inplace of `MusicBrainz ID`.
> 
> `Deep-Diff` isn't working properly. Should it be replaced w some different library?

Same as before, then. That's unfortunate.
I wonder if there's a way to sort the items by id before doing the comparison, and whether that would help.

Would it help to print out the data of two specific revisions (before/after) to be able to test it in a sandbox?

prabalsingh24 commented:
We ARE sorting it by ids. https://github.com/bookbrainz/bookbrainz-data-js/blob/1b3bf5c12326a8419401f283e25e014aca917347/src/util.js#L158

how about we compare relationshipSet and IdentifierSet by brute force. Picking up one relationship from the old set and check if it's present in the new set? Because we are facing problem only in sets (in the array). Lemme try this once. I think I can get it done.

MonkeyDo commented:
> We ARE sorting it by ids. https://github.com/bookbrainz/bookbrainz-data-js/blob/1b3bf5c12326a8419401f283e25e014aca917347/src/util.js#L158
> 
> how about we compare relationshipSet and IdentifierSet by brute force. Picking up one relationship from the old set and check if it's present in the new set? Because we are facing problem only in sets (in the array). Lemme try this once. I think I can get it done.

Hmmmm. It's been a while I looked into this in detail :p
I wonder then if on the contrary you should try not sorting the array and letting the new library handle it.
That being said, I don't know if we can assume the sets will be returned in the same order every time by the ORM…

MonkeyDo commented:
Looking at [the sorting code](https://github.com/bookbrainz/bookbrainz-data-js/blob/1b3bf5c12326a8419401f283e25e014aca917347/src/util.js#L124) again, it looks like we're not sorting by ID but rather by value and type.label, which might explain why it's all over the place.
You could try sorting by id or by type.id, that might do the trick.

prabalsingh24 commented:
@MonkeyDo I'll be afk for about a week. I'll continue this later. Is that alright?

MonkeyDo commented:
> @MonkeyDo I'll be afk for about a week. I'll continue this later. Is that alright?

Of course! You certainly do not need my permission !
Take care!

MonkeyDo commented on src/server/routes/revision.js:
```diff
@@ -16,7 +16,7 @@
  * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
  */
-
+import * as auth from '../helpers/auth';
```
This is not used

MonkeyDo commented on src/server/routes/revision.js:
```diff
@@ -169,12 +169,14 @@ function diffRevisionsWithParents(revisions) {
 						changes: revision.diff(parent),
-						entity: revision.related('entity')
+						entity: revision.related('entity'),
+						entityAlias: revision.related('data').fetch({require: false, withRelated: ['aliasSet.defaultAlias']})
```
Instead of fetching here, you can add `'data.aliasSet.defaultAlias'` to withRelated in `_createRevision` and just use `revision.related('data')` as we do for `entity` just above.

Should be replicated for the identical promise below.

MonkeyDo commented:
You can Ignore the previous review.
The rel formatting issue I was seeing was due to my local copy of bookbrainz-data that I was doing experiments with, didn't catch that at the time.
As for the suggestion to fetch 'data.aliasSet.defaultAlias', that doesn't seem to work either.
